File: sagco-os/sagco/lang/parser.py
"""
SAGCO language recursive descent parser.
Parses SAGCO token stream into a Program AST.
"""
from .tokens import Token, TokenType
from .ast_nodes import (
    Program, CellAssign, VarAssign, AssertStmt, PrintStmt,
    IfStmt, ForStmt, ReturnStmt, FuncCall, BinOp, UnaryOp,
    CellRef, Ident, Literal
)


class ParseError(SyntaxError):
    pass


class Parser:
    def __init__(self, tokens: list):
        # Filter out NEWLINE tokens for simpler parsing (we handle them as statement separators)
        self.tokens = [t for t in tokens if t.type != TokenType.NEWLINE or True]
        self.pos = 0

    def peek(self) -> Token:
        while self.pos < len(self.tokens) and self.tokens[self.pos].type == TokenType.NEWLINE:
            self.pos += 1
        if self.pos < len(self.tokens):
            return self.tokens[self.pos]
        return Token(TokenType.EOF, "", 0, 0)

    def peek_raw(self) -> Token:
        if self.pos < len(self.tokens):
            return self.tokens[self.pos]
        return Token(TokenType.EOF, "", 0, 0)

    def advance(self) -> Token:
        tok = self.tokens[self.pos]
        self.pos += 1
        return tok

    def skip_newlines(self):
        while self.pos < len(self.tokens) and self.tokens[self.pos].type == TokenType.NEWLINE:
            self.pos += 1

    def expect(self, ttype: TokenType) -> Token:
        self.skip_newlines()
        tok = self.tokens[self.pos] if self.pos < len(self.tokens) else Token(TokenType.EOF, "", 0, 0)
        if tok.type != ttype:
            raise ParseError(
                f"Expected {ttype.name} but got {tok.type.name} ({tok.value!r}) at line {tok.line}"
            )
        self.pos += 1
        return tok

    def check(self, *ttypes: TokenType) -> bool:
        self.skip_newlines()
        if self.pos >= len(self.tokens):
            return False
        return self.tokens[self.pos].type in ttypes

    def match(self, *ttypes: TokenType) -> Token | None:
        self.skip_newlines()
        if self.pos < len(self.tokens) and self.tokens[self.pos].type in ttypes:
            return self.advance()
        return None

    def parse(self) -> Program:
        stmts = []
        while not self.check(TokenType.EOF):
            self.skip_newlines()
            if self.check(TokenType.EOF):
                break
            stmt = self.parse_statement()
            if stmt is not None:
                stmts.append(stmt)
        return Program(stmts)

    def parse_statement(self):
        tok = self.peek()
        if tok.type == TokenType.CELL:
            return self.parse_cell_assign()
        elif tok.type == TokenType.ASSERT:
            return self.parse_assert()
        elif tok.type == TokenType.PRINT:
            return self.parse_print()
        elif tok.type == TokenType.IF:
            return self.parse_if()
        elif tok.type == TokenType.FOR:
            return self.parse_for()
        elif tok.type == TokenType.RETURN:
            return self.parse_return()
        elif tok.type == TokenType.IDENT:
            # Could be variable assignment: IDENT = expr
            return self.parse_ident_stmt()
        elif tok.type == TokenType.NEWLINE:
            self.advance()
            return None
        elif tok.type == TokenType.EOF:
            return None
        else:
            # Unknown token — skip the line
            self._skip_line()
            return None

    def _skip_line(self):
        """Skip tokens until NEWLINE or EOF."""
        while self.pos < len(self.tokens):
            t = self.tokens[self.pos]
            if t.type in (TokenType.NEWLINE, TokenType.EOF):
                break
            self.pos += 1

    def parse_cell_assign(self) -> CellAssign:
        self.expect(TokenType.CELL)
        ref_tok = self.expect(TokenType.CELLREF)
        self.expect(TokenType.ASSIGN)
        expr = self.parse_expr()
        return CellAssign(ref_tok.value, expr)

    def parse_assert(self) -> AssertStmt:
        self.expect(TokenType.ASSERT)
        # Parse the left side without consuming comparison operators
        left = self.parse_additive()
        op_tok = self.match(TokenType.EQ, TokenType.NEQ, TokenType.LT,
                            TokenType.GT, TokenType.LTEQ, TokenType.GTEQ)
        if op_tok is None:
            # Check if left is already a BinOp comparison (parse_expr consumed it)
            # Wrap as assert True
            return AssertStmt(left, "==", Literal(True))
        right = self.parse_additive()
        return AssertStmt(left, op_tok.value, right)

    def parse_print(self) -> PrintStmt:
        self.expect(TokenType.PRINT)
        expr = self.parse_expr()
        return PrintStmt(expr)

    def parse_if(self) -> IfStmt:
        self.expect(TokenType.IF)
        cond = self.parse_expr()
        self.match(TokenType.THEN)  # optional THEN keyword
        self.skip_newlines()
        then_stmts = []
        else_stmts = []
        while not self.check(TokenType.ELSE, TokenType.END, TokenType.EOF):
            self.skip_newlines()
            if self.check(TokenType.ELSE, TokenType.END, TokenType.EOF):
                break
            stmt = self.parse_statement()
            if stmt is not None:
                then_stmts.append(stmt)
        if self.match(TokenType.ELSE):
            self.skip_newlines()
            while not self.check(TokenType.END, TokenType.EOF):
                self.skip_newlines()
                if self.check(TokenType.END, TokenType.EOF):
                    break
                stmt = self.parse_statement()
                if stmt is not None:
                    else_stmts.append(stmt)
        self.match(TokenType.END)
        return IfStmt(cond, then_stmts, else_stmts)

    def parse_for(self) -> ForStmt:
        self.expect(TokenType.FOR)
        var_tok = self.expect(TokenType.IDENT)
        self.expect(TokenType.IN)
        iterable = self.parse_expr()
        self.skip_newlines()
        body = []
        while not self.check(TokenType.END, TokenType.EOF):
            self.skip_newlines()
            if self.check(TokenType.END, TokenType.EOF):
                break
            stmt = self.parse_statement()
            if stmt is not None:
                body.append(stmt)
        self.match(TokenType.END)
        return ForStmt(var_tok.value, iterable, body)

    def parse_return(self) -> ReturnStmt:
        self.expect(TokenType.RETURN)
        expr = self.parse_expr()
        return ReturnStmt(expr)

    def parse_ident_stmt(self):
        """Handle IDENT = expr (variable assignment) or standalone expression."""
        self.skip_newlines()
        ident_tok = self.tokens[self.pos]
        # Peek ahead to see if next non-nl is ASSIGN
        saved = self.pos
        self.pos += 1
        self.skip_newlines()
        if self.pos < len(self.tokens) and self.tokens[self.pos].type == TokenType.ASSIGN:
            self.pos += 1  # consume =
            expr = self.parse_expr()
            return VarAssign(ident_tok.value, expr)
        else:
            self.pos = saved
            # Just evaluate and discard (expression statement)
            expr = self.parse_expr()
            return PrintStmt(expr)  # treat as print

    # ---- Expression parsing (precedence climbing) ----

    def parse_expr(self):
        return self.parse_comparison()

    def parse_comparison(self):
        left = self.parse_additive()
        while True:
            op = self.match(TokenType.EQ, TokenType.NEQ, TokenType.LT,
                            TokenType.GT, TokenType.LTEQ, TokenType.GTEQ)
            if op is None:
                break
            right = self.parse_additive()
            left = BinOp(left, op.value, right)
        return left

    def parse_additive(self):
        left = self.parse_multiplicative()
        while True:
            op = self.match(TokenType.PLUS, TokenType.MINUS)
            if op is None:
                break
            right = self.parse_multiplicative()
            left = BinOp(left, op.value, right)
        return left

    def parse_multiplicative(self):
        left = self.parse_unary()
        while True:
            op = self.match(TokenType.STAR, TokenType.SLASH)
            if op is None:
                break
            right = self.parse_unary()
            left = BinOp(left, op.value, right)
        return left

    def parse_unary(self):
        self.skip_newlines()
        op = self.match(TokenType.MINUS)
        if op:
            operand = self.parse_primary()
            return UnaryOp("-", operand)
        return self.parse_primary()

    def parse_primary(self):
        self.skip_newlines()
        if self.pos >= len(self.tokens):
            return Literal(None)
        tok = self.tokens[self.pos]

        if tok.type == TokenType.NUMBER:
            self.pos += 1
            try:
                val = float(tok.value)
                if val == int(val) and "." not in tok.value:
                    val = int(val)
            except ValueError:
                val = 0
            return Literal(val)

        if tok.type == TokenType.STRING:
            self.pos += 1
            return Literal(tok.value)

        if tok.type == TokenType.BOOL:
            self.pos += 1
            return Literal(tok.value.upper() == "TRUE")

        if tok.type == TokenType.NULL:
            self.pos += 1
            return Literal(None)

        if tok.type == TokenType.CELLREF:
            self.pos += 1
            return CellRef(tok.value)

        if tok.type == TokenType.IDENT:
            self.pos += 1
            # Check if function call
            self.skip_newlines()
            if self.pos < len(self.tokens) and self.tokens[self.pos].type == TokenType.LPAREN:
                return self.parse_func_call(tok.value)
            return Ident(tok.value)

        if tok.type == TokenType.LPAREN:
            self.pos += 1
            expr = self.parse_expr()
            self.expect(TokenType.RPAREN)
            return expr

        # Unknown — skip and return None literal
        self.pos += 1
        return Literal(None)

    def parse_func_call(self, name: str) -> FuncCall:
        self.expect(TokenType.LPAREN)
        args = []
        if not self.check(TokenType.RPAREN):
            args.append(self.parse_expr())
            while self.match(TokenType.COMMA):
                if self.check(TokenType.RPAREN):
                    break
                args.append(self.parse_expr())
        self.expect(TokenType.RPAREN)
        return FuncCall(name, args)

File: BOARD-36-SAGCO-FORMULA-COMPILER/src/parser.rs
use crate::ast::{BinOp, Expr, UnaryOp};
use crate::token::{Spanned, Token};
use thiserror::Error;

#[derive(Debug, Error)]
pub enum ParseError {
    #[error("Unexpected token {0:?} at line {1}, col {2}")]
    UnexpectedToken(Token, usize, usize),
    #[error("Expected {0}, got {1:?}")]
    Expected(String, Token),
    #[error("Unexpected end of input")]
    UnexpectedEof,
}

pub struct Parser {
    tokens: Vec<Spanned<Token>>,
    pos: usize,
}

impl Parser {
    pub fn new(tokens: Vec<Spanned<Token>>) -> Self {
        Parser { tokens, pos: 0 }
    }

    fn peek(&self) -> &Token {
        self.tokens
            .get(self.pos)
            .map(|s| &s.token)
            .unwrap_or(&Token::EOF)
    }

    fn peek_span(&self) -> (usize, usize) {
        self.tokens
            .get(self.pos)
            .map(|s| (s.span.line, s.span.col))
            .unwrap_or((0, 0))
    }

    fn advance(&mut self) -> &Token {
        let tok = &self.tokens[self.pos].token;
        if self.pos < self.tokens.len() - 1 {
            self.pos += 1;
        }
        tok
    }

    fn expect(&mut self, expected: Token) -> Result<(), ParseError> {
        if self.peek() == &expected {
            self.advance();
            Ok(())
        } else {
            Err(ParseError::Expected(
                format!("{:?}", expected),
                self.peek().clone(),
            ))
        }
    }

    pub fn parse(&mut self) -> Result<Expr, ParseError> {
        let expr = self.parse_or()?;
        Ok(expr)
    }

    fn parse_or(&mut self) -> Result<Expr, ParseError> {
        let mut left = self.parse_and()?;
        loop {
            match self.peek() {
                Token::Ident(s) if s == "OR" => {
                    self.advance();
                    let right = self.parse_and()?;
                    left = Expr::BinOp {
                        left: Box::new(left),
                        op: BinOp::Or,
                        right: Box::new(right),
                    };
                }
                _ => break,
            }
        }
        Ok(left)
    }

    fn parse_and(&mut self) -> Result<Expr, ParseError> {
        let mut left = self.parse_comparison()?;
        loop {
            match self.peek() {
                Token::Ident(s) if s == "AND" => {
                    self.advance();
                    let right = self.parse_comparison()?;
                    left = Expr::BinOp {
                        left: Box::new(left),
                        op: BinOp::And,
                        right: Box::new(right),
                    };
                }
                _ => break,
            }
        }
        Ok(left)
    }

    fn parse_comparison(&mut self) -> Result<Expr, ParseError> {
        let mut left = self.parse_concat()?;
        loop {
            let op = match self.peek() {
                Token::Eq => BinOp::Eq,
                Token::NEq => BinOp::NEq,
                Token::Lt => BinOp::Lt,
                Token::Gt => BinOp::Gt,
                Token::LtEq => BinOp::LtEq,
                Token::GtEq => BinOp::GtEq,
                _ => break,
            };
            self.advance();
            let right = self.parse_concat()?;
            left = Expr::BinOp {
                left: Box::new(left),
                op,
                right: Box::new(right),
            };
        }
        Ok(left)
    }

    fn parse_concat(&mut self) -> Result<Expr, ParseError> {
        let mut left = self.parse_additive()?;
        loop {
            if self.peek() == &Token::Ampersand {
                self.advance();
                let right = self.parse_additive()?;
                left = Expr::BinOp {
                    left: Box::new(left),
                    op: BinOp::Concat,
                    right: Box::new(right),
                };
            } else {
                break;
            }
        }
        Ok(left)
    }

    fn parse_additive(&mut self) -> Result<Expr, ParseError> {
        let mut left = self.parse_multiplicative()?;
        loop {
            let op = match self.peek() {
                Token::Plus => BinOp::Add,
                Token::Minus => BinOp::Sub,
                _ => break,
            };
            self.advance();
            let right = self.parse_multiplicative()?;
            left = Expr::BinOp {
                left: Box::new(left),
                op,
                right: Box::new(right),
            };
        }
        Ok(left)
    }

    fn parse_multiplicative(&mut self) -> Result<Expr, ParseError> {
        let mut left = self.parse_power()?;
        loop {
            let op = match self.peek() {
                Token::Star => BinOp::Mul,
                Token::Slash => BinOp::Div,
                _ => break,
            };
            self.advance();
            let right = self.parse_power()?;
            left = Expr::BinOp {
                left: Box::new(left),
                op,
                right: Box::new(right),
            };
        }
        Ok(left)
    }

    fn parse_power(&mut self) -> Result<Expr, ParseError> {
        let base = self.parse_unary()?;
        if self.peek() == &Token::Caret {
            self.advance();
            let exp = self.parse_power()?; // right-associative
            Ok(Expr::BinOp {
                left: Box::new(base),
                op: BinOp::Pow,
                right: Box::new(exp),
            })
        } else {
            Ok(base)
        }
    }

    fn parse_unary(&mut self) -> Result<Expr, ParseError> {
        match self.peek().clone() {
            Token::Minus => {
                self.advance();
                let expr = self.parse_unary()?;
                Ok(Expr::UnaryOp {
                    op: UnaryOp::Neg,
                    expr: Box::new(expr),
                })
            }
            Token::Bang => {
                self.advance();
                let expr = self.parse_unary()?;
                Ok(Expr::UnaryOp {
                    op: UnaryOp::Not,
                    expr: Box::new(expr),
                })
            }
            _ => self.parse_primary(),
        }
    }

    fn parse_primary(&mut self) -> Result<Expr, ParseError> {
        match self.peek().clone() {
            Token::Number(n) => {
                self.advance();
                Ok(Expr::Number(n))
            }
            Token::Str(s) => {
                self.advance();
                Ok(Expr::Str(s))
            }
            Token::Bool(b) => {
                self.advance();
                Ok(Expr::Bool(b))
            }
            Token::CellRef(addr) => {
                self.advance();
                Ok(Expr::CellRef(addr))
            }
            Token::RangeRef(from, to) => {
                self.advance();
                Ok(Expr::RangeRef(from, to))
            }
            Token::LParen => {
                self.advance();
                let expr = self.parse()?;
                self.expect(Token::RParen)?;
                Ok(expr)
            }
            Token::Ident(name) => {
                self.advance();
                // Special IF form
                if name == "IF" {
                    return self.parse_if();
                }
                // Function call
                if self.peek() == &Token::LParen {
                    self.advance(); // consume '('
                    let args = self.parse_args()?;
                    self.expect(Token::RParen)?;
                    Ok(Expr::Call { name, args })
                } else {
                    // bare identifier (shouldn't normally happen but handle gracefully)
                    Ok(Expr::Str(name))
                }
            }
            Token::EOF => Err(ParseError::UnexpectedEof),
            other => {
                let (line, col) = self.peek_span();
                Err(ParseError::UnexpectedToken(other, line, col))
            }
        }
    }

    fn parse_if(&mut self) -> Result<Expr, ParseError> {
        self.expect(Token::LParen)?;
        let cond = self.parse()?;
        // accept comma or semicolon
        match self.peek() {
            Token::Comma | Token::Semicolon => { self.advance(); }
            _ => return Err(ParseError::Expected("comma".to_string(), self.peek().clone())),
        }
        let then = self.parse()?;
        match self.peek() {
            Token::Comma | Token::Semicolon => { self.advance(); }
            _ => return Err(ParseError::Expected("comma".to_string(), self.peek().clone())),
        }
        let else_ = self.parse()?;
        self.expect(Token::RParen)?;
        Ok(Expr::IfExpr {
            cond: Box::new(cond),
            then: Box::new(then),
            else_: Box::new(else_),
        })
    }

    fn parse_args(&mut self) -> Result<Vec<Expr>, ParseError> {
        let mut args = Vec::new();
        if self.peek() == &Token::RParen {
            return Ok(args);
        }
        args.push(self.parse()?);
        loop {
            match self.peek() {
                Token::Comma | Token::Semicolon => {
                    self.advance();
                    if self.peek() == &Token::RParen {
                        break;
                    }
                    args.push(self.parse()?);
                }
                _ => break,
            }
        }
        Ok(args)
    }
}

pub fn parse(input: &str) -> Result<Expr, ParseError> {
    use crate::lexer::tokenize;

    let tokens = tokenize(input).map_err(|e| ParseError::Expected(e.to_string(), Token::EOF))?;
    let mut parser = Parser::new(tokens);
    parser.parse()
}
